### ISSUE TITLE
[MNT] Duplicate MatrixProfile transformers? Why are there two?

### ISSUE THREAD

**Issue by MatthewMiddlehurst** [sktime/sktime#2347]
There are two matrix_profile.py files in the transformations package, one in series and one in panel. Are these the same transform, or are they doing different things using the matrix profile? I think the panel one at least is using the matrix profile as a distance (MPDist), would this not be better as a distance rather than a transformer?

**Comment by fkiraly** [comment]
I know the GitHub ID of @mloning, but do you know that of Claudia Sanchez who did not leave her GitHub ID in the code?

**Comment by mloning** [comment]
If I remember correctly, one is a custom implementation for time series classification, the other one uses stumpy.

**Comment by TonyBagnall** [comment]
Claudia was an intern of mine a while back, I agree this needs looking at

**Comment by TonyBagnall** [comment]
I'm guessing it will be on the contributor list. Also adding sorting out MPDist to the to do list here, 
https://github.com/alan-turing-institute/sktime/blob/main/sktime/distances/mpdist.py
it needs tidying up and converting to numba

**Comment by seanlaw** [comment]
In case it may be relevant to this group, in addition to computing a matrix profile (`stumpy.stump`), STUMPY also provides a [`stumpy.mpdist` function](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.mpdist), which leverages `numba` out of the box. All of the code is unit tested and both `stump` and `mpdist` functions are parallelized to use all available cores, which means that it will be significantly faster than the current `numpy` implementation and will scale to much longer time series.

**Comment by fkiraly** [comment]
@seanlaw, the `MatrixProfileTransformer` already uses `stumpy` - so, can you explain why you are providing this information, or how it is relevant to the above? I may just be too stupid to see the connection, so feel free to explain it like to an elementary school pupil.

**Comment by fkiraly** [comment]
are you referring, perchance, to the `mpdist` in `sktime`?

**Comment by seanlaw** [comment]
@fkiraly Yes, my apologies for not being so clear. My comment was in response to @TonyBagnall but also, in general, that you may be able to replace all of the existing matrix profile code with STUMPY and avoid needing to maintain that code in `sktime`. Sorry for the confusion

**Comment by fkiraly** [comment]
>in general, that you may be able to replace all of the existing matrix profile code with STUMPY and avoid needing to maintain that code in sktime. Sorry for the confusion

I see!

While you are here, do you perhaps understand the difference between the two matrix profile transformers in sktime?
I.e., why does one produce a series, and one non-time-ordered features? What are the outputs?

Alternatively, what is the matrix profile for you: an algorithm that produces a time series given a time series, or an algorithm that produces an unordered collection of features given a time series?

**Comment by seanlaw** [comment]
> While you are here, @seanlaw, do you perhaps understand the difference between the two matrix profile transformers in sktime? I.e., why does one produce a series, and one non-time-ordered features? What are the outputs?

> Alternatively, what is the matrix profile for you: an algorithm that produces a time series or sequence given a time series, or an algorithm that produces an unordered collection of features given a time series?

@fkiraly When I look at [series/matrix_profile.py](https://github.com/alan-turing-institute/sktime/blob/main/sktime/transformations/series/matrix_profile.py) and [panel/matrix_profile.py](https://github.com/alan-turing-institute/sktime/blob/main/sktime/transformations/panel/matrix_profile.py), for all intents and purposes, they both produce the same matrix profile output, which is a truncated 1D series that can be used to annotate the original input time series. Inherently, the matrix profile must be ordered and is **not** an unordered collection of features.

**Comment by fkiraly** [comment]
I see, thanks, @seanlaw!

So, can I confirm, you are saying:

* content-wise, both transformers look and seem equivalent, except that one formats the output as a table and one as a series
* conceptually, what *should* happen is formatting as a series, since there is an inherent ordering in the matrix profile
* so, the one in "panel" is redundant since it could be obtained by taking the one in "series" and then reformatting the output as a table, from a series.

**Comment by seanlaw** [comment]
> content-wise, both transformers look and seem equivalent, except that one formats the output as a table and one as a series

To be precise, I believe that the one in "series" actually returns a 1-D `numpy` array (in case we are referring to `pandas` Series)

> conceptually, what should happen is formatting as a series, since there is an inherent ordering in the matrix profile

Yes

> so, the one in "panel" is redundant since it could be obtained by taking the one in "series" and then reformatting the output as a table, from a series.

Yes

> implementation-wise, both are different, the "series" one uses stumpy whereas the "panel" one is a native implementation. You think the "series" one has the better implementation, and you are absolutely unbiased in this opinion 😄

I genuinely hope that I'm not coming across as having an ego as I do not want to diminish the wonderful open source contributions of others or to make this a competition. Having spent a lot of time implementing a hyper-optimized matrix profile calculator, I want to acknowledge that this stuff is not trivial and sincerely applaud anybody who is willing to try. We'll only get better as a community by working together! 

TL;DR. - If `sktime` mainly deals with short time series (i.e., n < 5K) then the native single-threaded implementation is likely sufficient. For longer time series, you'll need better algorithms (available in STUMPY) and more hardware.

To state things factually, I think that the native implementation is good but, from my experience implementing a similar pure-numpy version in the past, this doesn't scale to time series that are beyond n = 5-10K+ data points in length due to the fact that the computational complexity of a matrix profile calculation is, at best, O(n^2) and the native implementation is likely not multithreaded. With STUMPY, you not only get a parallelized multi-threaded `numba` implementation out of the box but it also leverages the newer algorithm (faster) that was published recently, which allows scaling to about 1M+ data points on a local server. Beyond that time series length, STUMPY has first class support for `dask` as well as GPUs. A severely outdated performance chart can be found [here](https://github.com/TDAmeritrade/stumpy#performance). Finally, as an added bonus, beyond the basic matrix profile computation, you get the [full suite of tools](https://stumpy.readthedocs.io/en/latest/api.html) that are available in STUMPY including `mpdist`.

**Comment by fkiraly** [comment]
resolved by explanation. `sktime` policy allows multiple implementations of same algorithm to coexist, so no further action necessary.
